### PR TITLE
Lambda load old data

### PR DESCRIPTION
**Pull request by HarvindG**
- Created initial draft of lambda function that selects data from RDS that is from past 24hrs
- Created Dockerfile to dockerise lambda function
- Written shell script that installs FreeTDS (working on this still - not working yet)
- created temporary data sub-directory that stores temporary files whilst working with them

### CODE REVIEW

File: lambda-load-old-data/install_sql.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+
+/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```
Comment (shivani-63) [inline lambda-load-old-data/install_sql.sh]: Nice! You found a way to incorporate this, did you test running the Dockerfile to see if it works?

File: lambda-load-old-data/lambda_load_s3.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+from os import environ, path
+
+from boto3 import client
+from botocore.exceptions import ClientError
+from dotenv import load_dotenv
+from sqlalchemy import create_engine, sql, engine
+
+
+def get_database_connection() -> engine.base.Connection:
+    """Get a connection to the short term database"""
```
Comment (shivani-63) [inline lambda-load-old-data/lambda_load_s3.py]:
```suggestion
    """Get a connection to the short term database."""
```

File: lambda-load-old-data/lambda_load_s3.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import csv
+from os import environ, path
+
+from boto3 import client
+from botocore.exceptions import ClientError
+from dotenv import load_dotenv
+from sqlalchemy import create_engine, sql, engine
+
+
+def get_database_connection() -> engine.base.Connection:
```
Comment (shivani-63) [inline lambda-load-old-data/lambda_load_s3.py]: Maybe you can pass in config as an argument to this and the connecting to the client function, similar to what Zander suggested yesterday? (Just for stylistic purposes :) )

File: lambda-load-old-data/lambda_load_s3.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    sql_engine = create_engine(
+        f"mssql+pymssql://{environ['DB_USER']}:{environ['DB_PASSWORD']}@{environ['DB_HOST']}/?charset=utf8")
+
+    connection = sql_engine.connect()
+
+    return connection
+
+
+def get_s3_client() -> client:
+    """Get a connection to the relevant S3 bucket"""
```
Comment (shivani-63) [inline lambda-load-old-data/lambda_load_s3.py]:
```suggestion
    """Get a connection to the relevant S3 bucket."""
```

File: lambda-load-old-data/lambda_load_s3.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    s3_client = client("s3",
+                       aws_access_key_id=environ["AWS_ACCESS_KEY_ID"],
+                       aws_secret_access_key=environ["AWS_SECRET_ACCESS_KEY"])
+    return s3_client
+
+
+def extract_old_data_from_database(connection: engine.base.Connection) -> list[tuple]:
+    """Extract data older than a day from short-term-memory as a list"""
```
Comment (shivani-63) [inline lambda-load-old-data/lambda_load_s3.py]:
```suggestion
    """Extract data older than a day from short-term-memory as a list."""
```

File: lambda-load-old-data/lambda_load_s3.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                     WHERE recording_taken < GETDATE();
+                     """)
+
+    result = connection.execute(query).fetchall()
+
+    return result
+
+
+def write_list_of_tuples_to_csv(data: list[tuple], csv_filename: str = 'data/latest_data.csv') -> bool:
+    """Write a list of tuples to a CSV file"""
```
Comment (shivani-63) [inline lambda-load-old-data/lambda_load_s3.py]:
```suggestion
    """Write a list of tuples to a CSV file."""
```

File: lambda-load-old-data/lambda_load_s3.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    with open(csv_filename, 'w', newline='') as csvfile:
+        csv_writer = csv.writer(csvfile)
+
+        # Assuming the first tuple in the list contains column headers
+        if data:
+            csv_writer.writerows(data)
+
+
+def get_master_archive_csv(s3_client: client, bucket: str, folder_name: str):
+    """Load archive csv file"""
```
Comment (shivani-63) [inline lambda-load-old-data/lambda_load_s3.py]:
```suggestion
    """Load archive csv file."""
```

File: lambda-load-old-data/lambda_load_s3.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    contents = s3_client.list_objects(Bucket=bucket)["Contents"]
+    all_keys = [o["Key"] for o in contents]
+
+    for k in all_keys:
+        if k.endswith('.csv'):
+            s3_client.download_file(bucket, k, f"{folder_name}/{k}")
+
+
+def combine_csv_files(source_csv: str, target_csv: str):
+    """Insert previous days data into archive csv"""
```
Comment (shivani-63) [inline lambda-load-old-data/lambda_load_s3.py]:
```suggestion
    """Insert previous days data into archive csv."""
```

File: lambda-load-old-data/lambda_load_s3.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        source_data = list(source_reader)
+
+    # Append data to the target CSV file
+    with open(target_csv, 'a', newline='', encoding='utf-8') as target_file:
+        target_writer = csv.writer(target_file)
+        target_writer.writerows(source_data)
+
+
+def load_combined_archive_data_to_s3(s3_client, file_name, bucket, object_name=None):
+    """Load old data as csv from short-term-memory into S3 then delete csv file"""
```
Comment (shivani-63) [inline lambda-load-old-data/lambda_load_s3.py]:
```suggestion
    """Load old data as csv from short-term-memory into S3 then delete csv file."""
```

File: lambda-load-old-data/lambda_load_s3.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+    try:
+        s3_client.upload_file(file_name, bucket, object_name)
+    except ClientError:
+        return False
+    return True
+
+
+def handler(event=None, context=None):
+    """Lambda handler function"""
```
Comment (shivani-63) [inline lambda-load-old-data/lambda_load_s3.py]:
```suggestion
    """Lambda handler function."""
```

File: lambda-load-old-data/lambda_load_s3.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    load_dotenv()
+
+    conn = get_database_connection()
+    s3_client = get_s3_client()
+
+    # Retrieve yesterdays data from short-term database
+    latest_recording_data = extract_old_data_from_database(conn)
+
+    # Write yesterdays daa to csv into LOCAL_OUTPUT_DIR
```
Comment (shivani-63) [inline lambda-load-old-data/lambda_load_s3.py]:
```suggestion
    # Write yesterdays data to csv into LOCAL_OUTPUT_DIR
```

File: lambda-load-old-data/lambda_load_s3.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        object_name = path.basename(file_name)
+
+    try:
+        s3_client.upload_file(file_name, bucket, object_name)
+    except ClientError:
+        return False
+    return True
+
+
+def handler(event=None, context=None):
```
Comment (shivani-63) [inline lambda-load-old-data/lambda_load_s3.py]: Really easy to follow and understand lambda function!

File: lambda-load-old-data/lambda_load_s3.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+                     loc.continent
+                     FROM s_delta.recording AS rec
+                     JOIN s_delta.plant AS plant ON rec.plant_id = plant.plant_id
+                     JOIN s_delta.botanist AS bot ON plant.botanist_id = bot.botanist_id
+                     JOIN s_delta.location AS loc ON plant.location_id = loc.location_id;
+                     """)
+
+DELETE_QUERY = sql.text("""DELETE FROM s_delta.recording;""")
+
+COLUMNS = {"recording_id": "Recording ID", "soil_moisture": "Soil Moisture",
```
Comment (HarvindG) [inline lambda-load-old-data/lambda_load_s3.py]: love these static variables - makes the whole script much more readable

File: lambda-load-old-data/lambda_load_s3.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    conn = get_database_connection(environ)
+    s3_client = get_s3_client(environ)
+
+    arch_data = get_archive_data_csv(s3_client, ARCHIVE_BUCKET, ARCHIVE_KEY)
+    new_plant_data = extract_old_data_from_database(conn)
+    combined_data = update_archive_data(arch_data, new_plant_data)
+
+    combined_data.to_csv(
+        f"s3://{ARCHIVE_BUCKET}/{ARCHIVE_KEY}", index=False
+    )
```
Comment (HarvindG) [inline lambda-load-old-data/lambda_load_s3.py]: So this updates straight to the S3 bucket? thats really neat didn't know you could do that - removes the need for an upload function

File: lambda-load-old-data/lambda_load_s3.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    """Updates the archived dataframe with new data."""
+
+    return pd.concat([arch_df, new_df], ignore_index=True, sort=False)
+
+
+def delete_data_from_db(conn: Connection) -> None:
+    """Deletes all data from the database."""
+
+    conn.execute(DELETE_QUERY)
+    conn.execute(sql.text("COMMIT;"))
```
Comment (CharliedeanCS) [inline lambda-load-old-data/lambda_load_s3.py]: Loving the small simple functions with the use of constants. Makes it very readable

File: ecs-load-old-data/ecs_load_to_s3.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+from io import StringIO
+from os import environ, _Environ
+
+import pandas as pd
+from boto3 import client
+from mypy_boto3_s3 import S3Client
+from dotenv import load_dotenv
+from sqlalchemy import create_engine, sql, Connection
+
+ARCHIVE_BUCKET = "c9-ladybird-lnhm-data-bucket"
```
Comment (shivani-63) [inline ecs-load-old-data/ecs_load_to_s3.py:15]: This is really nice, may be worth changing in some of the files I've worked on too to make it uniform :)

File: ecs-load-old-data/ecs_load_to_s3.py [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+    obj = s3_client.get_object(Bucket=bucket, Key=key)
+    csv_str = obj["Body"].read().decode()
+    return pd.read_csv(StringIO(csv_str))
+
+
+def update_archive_data(arch_df: pd.DataFrame, new_df: pd.DataFrame) -> pd.DataFrame:
+    """Updates the archived dataframe with new data."""
+
+    return pd.concat([arch_df, new_df], ignore_index=True, sort=False)
```
Comment (CharliedeanCS) [inline ecs-load-old-data/ecs_load_to_s3.py:76]: I like the use of concat to simply merge the dataframes

File: ecs-load-old-data/ecs_load_to_s3.py [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+
+    connection.execute(sql.text("USE plants;"))
+
+    result = connection.execute(GET_QUERY).fetchall()
+
+    return pd.DataFrame(result).rename(columns=COLUMNS)
+
+
+def get_archive_data_csv(s3_client: S3Client, bucket: str, key: str) -> pd.DataFrame:
+    """Retrieves the archived data from an S3 bucket."""
```
Comment (CharliedeanCS) [inline ecs-load-old-data/ecs_load_to_s3.py:66]: Nice use of doc strings throughout

**Review by CharliedeanCS — DISMISSED** [review]
Nice code and comments throughout.

**Review by CharliedeanCS — DISMISSED** [review]

**Review by shivani-63 — APPROVED** [review]

**Review by shivani-63 — DISMISSED** [review]

**Review by zandersnow14 — APPROVED** [review]

**Review by zandersnow14 — APPROVED** [review]

**Pull request closed without merging**